zone/worldserver.cpp

```diff
@@ -1888,189 +1888,434 @@ void WorldServer::HandleMessage(uint16 opcode, const EQ::Net::Packet &p)
+				if (client_raid->members[member_index].member && client_raid->members[member_index].member->IsClient()) {
+					auto raid_member = client_raid->members[member_index].member->CastToClient();
+					raid_member->SpellFinished(CZSC->spell_id, raid_member);
+				}
+			}
+		}
+		break;
+	}
+	case ServerOP_CZCastSpellGuild:
 	{
-		CZClientSignal_Struct* CZCS = (CZClientSignal_Struct*)pack->pBuffer;
-		Client* client = entity_list.GetClientByCharID(CZCS->charid);
+		CZCastSpellGuild_Struct* CZSC = (CZCastSpellGuild_Struct*) pack->pBuffer;
+		auto client_list = entity_list.GetClientList();
+		for (auto client : client_list) {
+			if (client.second->GuildID() > 0 && client.second->GuildID() == CZSC->guild_id) {
+				client.second->SpellFinished(CZSC->spell_id, client.second);
+			}
+		}
+		break;
+	}
+	case ServerOP_CZMarqueePlayer:
+	{
+		CZMarqueePlayer_Struct* CZMS = (CZMarqueePlayer_Struct*) pack->pBuffer;
+		auto client = entity_list.GetClientByCharID(CZMS->character_id);
+		std::string message = CZMS->message;
 		if (client != 0) {
-			client->Signal(CZCS->data);
+			client->SendMarqueeMessage(CZMS->type, CZMS->priority, CZMS->fade_in, CZMS->fade_out, CZMS->duration, message);
 		}
 		break;
 	}
-	case ServerOP_CZSignalGroup:
+	case ServerOP_CZMarqueeGroup:
 	{
-		CZGroupSignal_Struct* CZGS = (CZGroupSignal_Struct*)pack->pBuffer;
-		auto client_group = entity_list.GetGroupByID(CZGS->group_id);
+		CZMarqueeGroup_Struct* CZMS = (CZMarqueeGroup_Struct*) pack->pBuffer;
+		auto client_group = entity_list.GetGroupByID(CZMS->group_id);
+		std::string message = CZMS->message;
+		if (client_group) {			
+			for (int member_index = 0; member_index < MAX_GROUP_MEMBERS; member_index++) {
+				if (client_group->members[member_index] && client_group->members[member_index]->IsClient()) {
+					auto group_member = client_group->members[member_index]->CastToClient();
+					group_member->SendMarqueeMessage(CZMS->type, CZMS->priority, CZMS->fade_in, CZMS->fade_out, CZMS->duration, message);
+				}
+			}
+		}
+		break;
+	}
+	case ServerOP_CZMarqueeRaid:
+	{
+		CZMarqueeRaid_Struct* CZMS = (CZMarqueeRaid_Struct*) pack->pBuffer;
+		auto client_raid = entity_list.GetRaidByID(CZMS->raid_id);
+		std::string message = CZMS->message;
+		if (client_raid) {
+			for (int member_index = 0; member_index < MAX_RAID_MEMBERS; member_index++) {
+				if (client_raid->members[member_index].member && client_raid->members[member_index].member->IsClient()) {
+					auto raid_member = client_raid->members[member_index].member->CastToClient();
+					raid_member->SendMarqueeMessage(CZMS->type, CZMS->priority, CZMS->fade_in, CZMS->fade_out, CZMS->duration, message);
+				}
+			}
+		}
+		break;
+	}
+	case ServerOP_CZMarqueeGuild:
+	{
+		CZMarqueeGuild_Struct* CZMS = (CZMarqueeGuild_Struct*) pack->pBuffer;
+		auto client_list = entity_list.GetClientList();
+		std::string message = CZMS->message;
+		for (auto client : client_list) {			
```

Comment:
Everywhere where you are doing this; you need to use `&` as right now you are performing memory copies in these loops

`for (auto &client : client_list) {` will use iterate over the client object in the client list by reference instead of copying the client object for each iteration; this is a big performance implication

Note: search in the source for anywhere you see `for (auto &` and you'll see this referenced almost everywhere and should be done in most cases